static/js/unread.js

```diff
@@ -6,11 +6,100 @@ var unread = (function () {
 var exports = {};
 
 var unread_mentioned = new Dict();
-var unread_topics = new Dict({fold_case: true});
 var unread_privates = new Dict(); // indexed by user_ids_string like 5,7,9
 exports.suppress_unread_counts = true;
 exports.messages_read_in_narrow = false;
 
+exports.unread_topic_counter = (function () {
+    var self = {};
+
+    function str_dict() {
+        // Use this when keys are streams and topics.
+        return new Dict({fold_case: true});
+    }
+
+    function num_dict() {
+        // Use this for message ids.
+        return new Dict();
+    }
+
+    var unread_topics = str_dict(); // dict of stream -> topic -> msg id
+
+    self.clear = function () {
+        unread_topics = str_dict();
+    };
+
+    self.update = function (stream, subject, new_subject, msg_id) {
+        if (unread_topics.has(stream) &&
+            unread_topics.get(stream).has(subject) &&
+            unread_topics.get(stream).get(subject).get(msg_id)) {
+            // Move the unread subject count to the new subject
+            unread_topics.get(stream).get(subject).del(msg_id);
+            if (unread_topics.get(stream).get(subject).num_items() === 0) {
+                unread_topics.get(stream).del(subject);
+            }
+            unread_topics.get(stream).setdefault(new_subject, num_dict());
+            unread_topics.get(stream).get(new_subject).set(msg_id, true);
+        }
+    };
+
+    self.add = function (stream, subject, msg_id) {
+        unread_topics.setdefault(stream, str_dict());
+        unread_topics.get(stream).setdefault(subject, num_dict());
+        unread_topics.get(stream).get(subject).set(msg_id, true);
+    };
+
+
+    self.del = function (stream, subject, msg_id) {
+        var stream_dict = unread_topics.get(stream);
+        if (stream_dict) {
+            var subject_dict = stream_dict.get(subject);
+            if (subject_dict) {
+                subject_dict.del(msg_id);
+            }
+        }
+    };
+
+    self.get_counts = function (res) {
```

Comment:
Any suggestion for the name?  I think `set_counts` makes it sound too much like you're muting the internal data structures of the class instead of the the `res` object.  This is a case where it would be nice if JS had tuples support, and I'd return two values that the caller could put into its result object.  Maybe `set_result_counts`?